### DISCUSSION TITLE
Using Alloy model and collection without a view data binding

### DISCUSSION THREAD

**Discussion by beppo-ivel** [appcelerator/titanium_mobile#13556] [Q&A] [answered]
I create some models of type random in `Util.record` and want to print them after clicking on a label. But the collection is undefined.

views/index.xml
```
<Alloy>
	<Window class="container" onClose="cleanup" >
		<Label id="label" onClick="doClick">Hello, World</Label>
	</Window>
</Alloy>
```
lib/util.js
```
class Util{

  static record(e){
    let no = Math.random() * 10;
    let model = Alloy.createModel('random', {name : "test" + no, number: no});
    model.save();
  }

}
module.exports = Util;
```
models/random.js

```
exports.definition = {
    config: {
        "columns": {
            "name": "TEXT",
            "number": "NUMBER",
            "id": "INTEGER PRIMARY KEY AUTOINCREMENT"
        },
        "adapter": {
            "type": "sql",
            "collection_name": "random",
            "idAttribute": "id"
        }
    },
    extendModel: function(Model) {
      _.extend(Model.prototype, {
        constructor : function ( obj ) {
          Alloy.Model.apply(this, {number: obj.number,
                                   name: obj.name});
        }
      });
      return Model;
    },
    extendCollection: function(Collection) {
      _.extend(Collection.prototype, { // Extend, override or implement Backbone.Collection
          clear: function() {
            _.invoke(this.toArray(), 'destroy');
          }
      });

      return Collection;
    }
}

```

controllers/index.js

```
let Util = require('/util');
function doClick(e) {
   Ti.Geolocation.removeEventListener('location', (e) => Util.record(e));
   let models = Alloy.Collections.random;
   Ti.API.error(JSON.stringify(models));

}

if (OS_ANDROID) {
	$.index.addEventListener('open', function() {
		if (!Ti.Geolocation.hasLocationPermissions()) {
	
		}else{
			Ti.Geolocation.accuracy = Ti.Geolocation.ACCURACY_HIGH;
   		        Ti.Geolocation.addEventListener('location', (e) => Util.record(e));
		}
	});
}
$.index.open();
function cleanup() {
    $.destroy();
}
```

**Comment by m1ga** [comment, marked as answer]
I like that you put an example in there but I've removed some parts to make it easier to debug/solve :smile: Feel free to add your location stuff back in (since that has nothing todo with the model I left it out)

Model and lib is the same.

```xml
<Alloy>
  <Collection src="random"/>
	<Window class="container" onClick="doClick" />
</Alloy>
```

```js
let Util = require('/util');

function doClick(e) {
	Util.record(e);
	let models = Alloy.Collections.random;
	models.fetch();
	console.log(JSON.stringify(models));
}


$.index.open();
```

click anywhere and it will output the model. I've registered the Collection in the XML file and do a `fetch()`.

**Reply by beppo-ivel** [reply]
> parts to make it easier to debug/solve

Sorry I will take care about next time.

1) If I wont register the model in the view, then I have to create it once (`Alloy.createCollection('random');`)? Is there a difference between both methods (e.g. the scope)? 

2) If I would not create or either register the the collection,  `Alloy.createModel(...); model.save()` creating and storing my objects anyway?

3) AFAIK `models.fetch();` is async, so maybe it will not finished before `console.log(JSON.stringify(models));` starts. Is that relevant for your example?

**Reply by m1ga** [reply]
> Sorry I will take care about next time.

no problem, I just want to minimize other mistakes/errors e.g. if there is a permission problem with the location or so :-)

You can look at the generated index.js in `/Resources/android/alloy/controllers/` and you can see that the Alloy `<Collection>` part generates `Alloy.Collections.instance('random');`

You can use
```js
Alloy.Collections.gyms.fetch({
	reset: true,
	success: function() {
		args.hideLoader();
		args.hideFilter();
	}
});
```
to get a success event. And I think you can even use `random.on("fetch", function(){})` but I don't have that in my code.

The rest I can't tell you. Would have to run tests myself. Maybe search for Backbone examples perhaps people solved those issues in normal JS already

**Reply by beppo-ivel** [reply]
> you can look at the generated index.js in `/Resources/android/alloy/controllers/ `

Thanks, got it.

With respect to my third question, which you answered in second: For this purpose I add  `myFetch` as Promise in `models/random.js`

```
myFetch: function(query){
            let self = this;
            return new Promise(function(resolve, reject) {
          	   self.fetch({query: query,
          		         success: function(collection, response, options){
          				   resolve(response);
          			 },
          		 	 error: function(collection, response, options){
                                          reject(response);
                                }
                        });
          	});
          }
```
